In [ ]:
import pandas as pd 
import matplotlib.pyplot as plt 
import numpy as np

import warnings
warnings.filterwarnings('ignore')

In [ ]:
df = pd.read_csv("https://archive.ics.uci.edu/ml/machine-learning-databases/00471/Data_for_UCI_named.csv")

In [ ]:
df.head()

tau1      tau2      tau3  ...        g4      stab     stabf
0  2.959060  3.079885  8.381025  ...  0.958034  0.055347  unstable
1  9.304097  4.902524  3.047541  ...  0.781760 -0.005957    stable
2  8.971707  8.848428  3.046479  ...  0.109853  0.003471  unstable
3  0.716415  7.669600  4.486641  ...  0.362718  0.028871  unstable
4  3.134112  7.608772  4.943759  ...  0.820923  0.049860  unstable

[5 rows x 14 columns]

In [ ]:
df.sample(5)

tau1      tau2      tau3  ...        g4      stab     stabf
7030  9.938786  3.475691  6.492763  ...  0.922036 -0.032816    stable
978   9.513874  6.772855  7.064591  ...  0.954858  0.055405  unstable
398   7.887232  5.835712  5.871384  ...  0.470315  0.045164  unstable
2948  6.358740  1.470855  8.471962  ...  0.189465  0.002365  unstable
567   8.642909  1.090504  9.943507  ...  0.732034 -0.010283    stable

[5 rows x 14 columns]

In [ ]:
#Dropping 'stab'
new_df=df.drop("stab",axis=1)
new_df

tau1      tau2      tau3  ...        g3        g4     stabf
0     2.959060  3.079885  8.381025  ...  0.887445  0.958034  unstable
1     9.304097  4.902524  3.047541  ...  0.562139  0.781760    stable
2     8.971707  8.848428  3.046479  ...  0.839444  0.109853  unstable
3     0.716415  7.669600  4.486641  ...  0.929381  0.362718  unstable
4     3.134112  7.608772  4.943759  ...  0.656947  0.820923  unstable
...        ...       ...       ...  ...       ...       ...       ...
9995  2.930406  9.487627  2.376523  ...  0.813512  0.608385  unstable
9996  3.392299  1.274827  2.954947  ...  0.285880  0.366120    stable
9997  2.364034  2.842030  8.776391  ...  0.149286  0.145984    stable
9998  9.631511  3.994398  2.757071  ...  0.889118  0.818391  unstable
9999  6.530527  6.781790  4.349695  ...  0.378761  0.942631  unstable

[10000 rows x 13 columns]

In [ ]:
#Checking null values
new_df.isnull().sum()

tau1     0
tau2     0
tau3     0
tau4     0
p1       0
p2       0
p3       0
p4       0
g1       0
g2       0
g3       0
g4       0
stabf    0
dtype: int64

In [ ]:
# Describing values
new_df.describe()

tau1          tau2  ...            g3            g4
count  10000.000000  10000.000000  ...  10000.000000  10000.000000
mean       5.250000      5.250001  ...      0.525000      0.525000
std        2.742548      2.742549  ...      0.274255      0.274255
min        0.500793      0.500141  ...      0.050054      0.050028
25%        2.874892      2.875140  ...      0.287514      0.287494
50%        5.250004      5.249981  ...      0.525015      0.525002
75%        7.624690      7.624893  ...      0.762440      0.762433
max        9.999469      9.999837  ...      0.999982      0.999930

[8 rows x 12 columns]

In [ ]:
# Spliting into test and train data
from sklearn.model_selection import train_test_split

In [ ]:
x_data = new_df.drop("stabf",axis=1)
y_data = new_df[["stabf"]]

In [ ]:
x_train,x_test,y_train,y_test = train_test_split(x_data,y_data,test_size=0.2,random_state=1)

In [ ]:
# Standard Scaler
from sklearn.preprocessing import StandardScaler

In [ ]:
SCALE = StandardScaler()

In [ ]:
SCALE.fit(x_data)

StandardScaler(copy=True, with_mean=True, with_std=True)

In [ ]:
X_train_Scaled=pd.DataFrame(SCALE.fit_transform(x_train), columns=x_train.columns)
X_test_Scaled=pd.DataFrame(SCALE.transform(x_test), columns=x_test.columns)


In [ ]:
X_train_Scaled.head()
print(X_train_Scaled.head())
print(X_test_Scaled.head())
y_train=y_train.reset_index(drop=True)

       tau1      tau2      tau3  ...        g2        g3        g4
0  0.367327 -0.986042  0.650447  ...  0.339859  0.585568  0.492239
1 -0.064659  0.089437  1.035079  ... -1.558488  1.429649 -1.443521
2 -1.467850  1.298418 -0.502536  ...  1.451534 -1.045743  0.492489
3  0.820081  0.529920  1.299657  ...  1.361958  1.604140  0.275303
4  0.665424 -1.425627  0.312300  ...  0.695660  1.137504 -1.312575

[5 rows x 12 columns]
       tau1      tau2      tau3  ...        g2        g3        g4
0  0.593951 -0.412733  1.503924  ...  1.167203 -1.507330  1.084726
1  0.202190  0.374416 -0.188800  ... -0.395660  1.414651  1.226011
2 -1.079044 -0.313745 -0.884634  ... -1.438495  0.651821 -1.682168
3 -0.083120 -1.107327  0.372805  ... -1.672322 -0.357714  1.055865
4  0.873921  1.438466  0.086662  ... -1.469731  0.956396 -0.819727

[5 rows x 12 columns]


In [ ]:
y_test=y_test.reset_index(drop=True)
print(y_test)

         stabf
0     unstable
1     unstable
2       stable
3       stable
4     unstable
...        ...
1995    stable
1996    stable
1997    stable
1998  unstable
1999  unstable

[2000 rows x 1 columns]


In [ ]:
from sklearn.preprocessing import LabelEncoder
encoder = LabelEncoder()
y_train =pd.DataFrame(encoder.fit_transform(y_train.reset_index(drop=True)), columns=y_train.columns)
y_test = pd.DataFrame(encoder.transform(y_test.reset_index(drop=True)),columns=y_test.columns) 
print(y_test)
y_train

      stabf
0         1
1         1
2         0
3         0
4         1
...     ...
1995      0
1996      0
1997      0
1998      1
1999      1

[2000 rows x 1 columns]


/usr/local/lib/python3.6/dist-packages/sklearn/preprocessing/_label.py:251: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.6/dist-packages/sklearn/preprocessing/_label.py:268: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


stabf
0         1
1         1
2         1
3         1
4         1
...     ...
7995      0
7996      0
7997      1
7998      1
7999      0

[8000 rows x 1 columns]

In [ ]:
# Training the Random forest classifier

In [ ]:
from sklearn.ensemble import RandomForestClassifier

In [ ]:
rf = RandomForestClassifier(n_estimators=100,random_state=1)
rf.fit(X_train_Scaled,y_train)

pred_y=rf.predict(X_test_Scaled)
pred_y

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:2: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  


array([1, 1, 0, ..., 0, 0, 1])

In [ ]:
# Statistical Report

from sklearn.metrics import classification_report,confusion_matrix,accuracy_score

In [ ]:
print(confusion_matrix(y_test,pred_y))
print(classification_report(y_test,pred_y))
print(accuracy_score(y_test, pred_y))

[[ 625   87]
 [  55 1233]]
              precision    recall  f1-score   support

           0       0.92      0.88      0.90       712
           1       0.93      0.96      0.95      1288

    accuracy                           0.93      2000
   macro avg       0.93      0.92      0.92      2000
weighted avg       0.93      0.93      0.93      2000

0.929


In [66]:
# XGBoost Classifier training

from xgboost import XGBClassifier

In [ ]:
model = XGBClassifier()
model.fit(X_train_Scaled, y_train)

/usr/local/lib/python3.6/dist-packages/sklearn/preprocessing/_label.py:235: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.6/dist-packages/sklearn/preprocessing/_label.py:268: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0,
              learning_rate=0.1, max_delta_step=0, max_depth=3,
              min_child_weight=1, missing=None, n_estimators=100, n_jobs=1,
              nthread=None, objective='binary:logistic', random_state=0,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
              silent=None, subsample=1, verbosity=1)

In [ ]:
1
print(model)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0,
              learning_rate=0.1, max_delta_step=0, max_depth=3,
              min_child_weight=1, missing=None, n_estimators=100, n_jobs=1,
              nthread=None, objective='binary:logistic', random_state=0,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
              silent=None, subsample=1, verbosity=1)


In [ ]:


# make predictions for test data
y_pred = model.predict(X_test_Scaled)
predictions = [round(value) for value in y_pred]


In [ ]:
#Accuracy
accuracy = accuracy_score(y_test, predictions)
print("Accuracy:",accuracy)

Accuracy: 0.9195


In [ ]:
# LGBM Classifier Training

In [ ]:
from lightgbm import LGBMClassifier

In [ ]:
model = LGBMClassifier()

In [ ]:
model.fit(X_train_Scaled,y_train)
pred_y=model.predict(X_test_Scaled)

/usr/local/lib/python3.6/dist-packages/sklearn/preprocessing/_label.py:235: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.6/dist-packages/sklearn/preprocessing/_label.py:268: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [ ]:
def score_data(y_test,pred_y):
  print("\n CONFUSION MATRIX \n", confusion_matrix(y_test,pred_y))
  print(classification_report(y_test,pred_y))
  print("Accuracy Score is: ", accuracy_score(y_test, pred_y))


score_data(y_test, pred_y)


 CONFUSION MATRIX 
 [[ 635   77]
 [  48 1240]]
              precision    recall  f1-score   support

           0       0.93      0.89      0.91       712
           1       0.94      0.96      0.95      1288

    accuracy                           0.94      2000
   macro avg       0.94      0.93      0.93      2000
weighted avg       0.94      0.94      0.94      2000

Accuracy Score is:  0.9375


In [ ]:
# Extra Trees Classifier Training

from sklearn.ensemble import ExtraTreesClassifier
#ExtraTreesClassifier
from sklearn.ensemble import ExtraTreesClassifier
ext_clf=ExtraTreesClassifier()
ext_clf.fit(X_train_Scaled,y_train)
pred_y=ext_clf.predict(X_test_Scaled)
score_data(y_test, pred_y)


 CONFUSION MATRIX 
 [[ 599  113]
 [  25 1263]]
              precision    recall  f1-score   support

           0       0.96      0.84      0.90       712
           1       0.92      0.98      0.95      1288

    accuracy                           0.93      2000
   macro avg       0.94      0.91      0.92      2000
weighted avg       0.93      0.93      0.93      2000

Accuracy Score is:  0.931


In [ ]:
#HyperTuning ExtraTreesClassifier
n_estimators = [50, 100, 300, 500, 1000]

min_samples_split = [2, 3, 5, 7, 9]

min_samples_leaf = [1, 2, 4, 6, 8]

max_features = ['auto', 'sqrt', 'log2', None] 

hyperparameter_grid = {'n_estimators': n_estimators,'min_samples_leaf': min_samples_leaf,'min_samples_split': min_samples_split,'max_features': max_features}

from sklearn.model_selection import RandomizedSearchCV
etc_clf=ExtraTreesClassifier()
rdm = RandomizedSearchCV(etc_clf, param_distributions = hyperparameter_grid, cv=5, n_iter=10, n_jobs=-1, verbose=1, random_state=1, scoring='accuracy')
rdm.fit(X_train_Scaled, y_train)
rdm.best_params_





Fitting 5 folds for each of 10 candidates, totalling 50 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 2 concurrent workers.
[Parallel(n_jobs=-1)]: Done  46 tasks      | elapsed:  1.5min
[Parallel(n_jobs=-1)]: Done  50 out of  50 | elapsed:  1.5min finished


{'max_features': None,
 'min_samples_leaf': 8,
 'min_samples_split': 2,
 'n_estimators': 1000}

In [ ]:
etc_clf_opt=ExtraTreesClassifier(max_features=None,min_samples_leaf=8,min_samples_split=2, n_estimators=1000)
etc_clf.fit(X_train_Scaled, y_train)
pred_y=etc_clf.predict(X_test_Scaled)
score_data(y_test, pred_y)


 CONFUSION MATRIX 
 [[ 606  106]
 [  32 1256]]
              precision    recall  f1-score   support

           0       0.95      0.85      0.90       712
           1       0.92      0.98      0.95      1288

    accuracy                           0.93      2000
   macro avg       0.94      0.91      0.92      2000
weighted avg       0.93      0.93      0.93      2000

Accuracy Score is:  0.931


The Accuracy for Optimal ExtraTrees Classifier is 0.928, whereas for Untuned ExtratreesClassifier is 0.931
Thus, Optimal Model has lower accuracy than the initial model

In [65]:
etc_clf_opt=ExtraTreesClassifier(max_features=None,min_samples_leaf=8,min_samples_split=2, n_estimators=1000)
etc_clf_opt.fit(X_train_Scaled, y_train)
pred_y=etc_clf_opt.predict(X_test_Scaled)
score_data(y_test, pred_y)
#Feature Importances for Optimal Model of ExtraTreesClassifier
feature_imp=etc_clf_opt.feature_importances_

xyz=pd.DataFrame(feature_imp, index=X_train_Scaled.columns, columns=['Importance'])
print("Max Importance is of: ", xyz[['Importance']].idxmax())
print("Min Importance is of: ", xyz[['Importance']].idxmin())


 CONFUSION MATRIX 
 [[ 622   90]
 [  54 1234]]
              precision    recall  f1-score   support

           0       0.92      0.87      0.90       712
           1       0.93      0.96      0.94      1288

    accuracy                           0.93      2000
   macro avg       0.93      0.92      0.92      2000
weighted avg       0.93      0.93      0.93      2000

Accuracy Score is:  0.928
Max Importance is of:  Importance    tau2
dtype: object
Min Importance is of:  Importance    p1
dtype: object
